### DISCUSSION TITLE
nan value in result

### DISCUSSION THREAD

**Discussion by clumytian** [ultralytics/ultralytics#10061] [Q&A]
![image](https://github.com/ultralytics/ultralytics/assets/85214730/f7cf3807-7e73-4ecc-a84a-71c831ad33d7)
I tried to use freeze layers. However, I get a lot of nan in precision, iou_loss, and cls_loss. Does it matter? Since I only focus on the mAP@50. 

One more thing, when I load the yolov8 models without loading yaml file, all of my layers' requires_grad is false. After I add yolov8.yaml, all of my layers' requires_gard is TRUE.
Although all the layer is false, I'm still able to train my data successfully. Does anyone know this why?

Previous when I do the CNN classification, I always freeze some layers during transfer learning. However, I found that freeze some layers might lower the accuracy when I use yolo and rtdert. Am I doing something wrong? 
Thanks a lot for helping me.
Best!

**Comment by pderrenger** [comment]
Hello there!

Observing `nan` in precision, iou_loss, and cls_loss might indicate that there's an issue such as exploding or vanishing gradients during your training process. While your focus might primarily be on mAP @clumytian, these `nan` values can signal that the model may not be learning effectively, which, in turn, could impact your targeted metric down the line. A good starting point to address this could be to adjust your learning rate or review your data normalization strategies. 🔄

Regarding the `requires_grad` behavior, it's expected! By default, when you load the model without a YAML file, it assumes you're in inference mode, hence `requires_grad` is `False` to optimize prediction speed. When you load a YAML (specifying training configurations), it sets up the model for training, making `requires_grad=True` for all parameters you're looking to update.

If your layers' `requires_grad` is `False` and you're still seeing successful training, it could be due to parts of the model still being trainable, or the metric improvements come from learning components other than the frozen layers. Freezing layers is advisable when doing transfer learning to retain learned features from another domain. However, the effectiveness depends greatly on the similarity between the pretraining and target tasks and might require fine-tuning for optimal performance.

Here's a brief example of adjusting the `requires_grad` attribute for freezing and unfreezing layers:

```python
from ultralytics import YOLO

# Load your model
model = YOLO('path/to/yolov8.pt')

# Freezing layers: setting requires_grad=False
for param in model.parameters():
    param.requires_grad = False

# Unfreezing a specific layer
for param in model.layer_to_train.parameters():
    param.requires_grad = True
```

In YOLO and other object detection tasks, the balance between retaining prelearned features and allowing for new feature learning specific to your dataset is crucial. Fine-tuning this balance for your specific use case may help avoid lowering accuracy.

Happy to help and wish you the best on your project! 😊
